### ISSUE TITLE
There's no way to handle validation errors with IOptionsMonitor.OnChange

### ISSUE THREAD

**Issue by davidfowl** [dotnet/runtime#44381]
Today you can use `IOptionsMonitor.OnChange` to subscribe to changes when an `IOptionsChangeTokenSource<T>` cancellation token source triggers. The problem is that if the next call results in an invalid configuration or throws an exception during the building of the changed configuration, that error will throw on a background thread and OnChange will never fire. Even worse is that there's no way to handle those validation errors or exceptions. This could also result in crashing the program if the `IOptionsChangeTokenSource<T>` implementation isn't handling exceptions properly.

Here's an example (change is never displayed on the screen):

```C#
using System;
using System.ComponentModel.DataAnnotations;
using System.Threading;
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.Options;
using Microsoft.Extensions.Primitives;

var services = new ServiceCollection();
services.AddOptions();

var num = 0;
services.AddOptions<MyOptions>()
        .Configure(o =>
        {
            num++;
            o.Name = num % 2 == 0 ? "David" : null;
        })
        .ValidateDataAnnotations();

var trigger = new TriggerChanges();
services.AddSingleton<IOptionsChangeTokenSource<MyOptions>>(trigger);

var sp = services.BuildServiceProvider();

var options = sp.GetRequiredService<IOptionsMonitor<MyOptions>>();

Console.WriteLine(options.CurrentValue);

options.OnChange(o =>
{
    Console.WriteLine("Changed!");
});

trigger.Change();

public class TriggerChanges : IOptionsChangeTokenSource<MyOptions>
{
    private CancellationTokenSource _cts = new CancellationTokenSource();

    public string Name => default;

    public void Change()
    {
        var old = _cts;
        _cts = new CancellationTokenSource();
        old.Cancel();
    }

    public IChangeToken GetChangeToken() => new CancellationChangeToken(_cts.Token);
}

public class MyOptions
{
    [Required]
    public string Name { get; set; }
    public int Age { get; set; }
}
```

I think the right fix for this is a design change around the OnChange callback where it doesn't provide you with the new options, but instead just fires and lets you read the new options lazily from the CurrentValue property. That would let you handle exceptions cleanly.

**Comment by davidfowl** [comment]
cc @HaoK did this come up during the initial design? 

I can think of 2 ways to address this:
- Implement *another* options interface with an `OnChange` callback that just notifies without first retrieving the new options
- Implement *another* options interface with an `OnChange` callback that pushes the error through (similar to Rx)

We may be able to retrofit it by using:
- Default interface methods in .NET 3/5
- Separate interface + extension method on .NET Standard 2.x

**Comment by HaoK** [comment]
Ah yeah, no we didn't really consider how validation errors would interact with OnChange

**Comment by davidfowl** [comment]
Yea, it doesn't work well right now if things throw on background threads. The good news is that this can be worked around today because all of the interfaces are public but it basically requires a re-implementation of the OptionsMonitor. I think we should make a tiny change here:

https://github.com/dotnet/runtime/blob/c82abc5cd40ecc9b61c6a8590acb267131727a98/src/libraries/Microsoft.Extensions.Options/src/OptionsMonitor.cs#L53

To only resolve the options if the callback is non-null. At least it would throw errors in the background unless you are interested in the changes. The only potential for breaking here is that errors maybe thrown from accesses to `CurrentValue`

**Comment by HaoK** [comment]
You prefer letting them fetch the options as opposed to pushing the error (with a null instance I guess to signify an error has occurred?

`IDisposable OnChange(Action<TOptions, string, Exception> listener);` where TOptions will be null whenever there's an exception.  The old OnChange could just call the new overload with an action that ignores the exception too.

**Comment by HaoK** [comment]
The tweak of moving the Get into the non null check I think would result in the cache not being populated which is a subtle (not sure if it matters difference)

**Comment by davidfowl** [comment]
@HaoK IDEA but likely breaking, what if we pass null with an emptied cache? That would allow you to handle the error by accessing the value but would still fire the callback...

**Comment by HaoK** [comment]
That seems decent, so basically we wrap the Get with a try/catch in OptionsMonitor and callers will need to guard against null (if they use options validation).

**Comment by HaoK** [comment]
Yeah maybe better for them to opt-in with a new OnChange than blow everyone up

**Comment by eerhardt** [comment]
6.0 is now feature complete. Moving to 7.0.

**Comment by xakep139** [comment]
@davidfowl are there any plans to address this issue in 9.0?

**Comment by tekian** [comment]
Yup, quite mission critical for services that use inherently unreliable source of configuration and/or feature flags with values that need validation on-the-fly. Maybe we could add an overload with an on-error semantics? 🫣

**Comment by geeknoid** [comment]
@danmoseley @joperezr Is there a chance this could be tackled in the next release?

**Comment by emiliovmq** [comment]
Is there currently any solution to support cases where there are validation errors (or any other error, such as the file does not exist, etc.) in the configuration, and using at the same time IOptionsMonitor and the OnChange event (to detect changes in the configuration source)? We cannot rely on this mechanism (which is also unique for singleton services, like ours) if there is no support for such case.

**Comment by BrandonSchreck** [comment]
Currently looking at implementing IOptionsMonitor and this came up...I'm running .NET 8 Isolated Azure Func and need to run validation when a change occurs to ensure dynamic specification pattern rules engine wont break.
